handler/src/main/java/io/netty/handler/timeout/WriteTimeoutHandler.java

```diff
@@ -111,31 +114,68 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
         ctx.write(msg, promise);
     }
 
-    private void scheduleTimeout(final ChannelHandlerContext ctx, final ChannelPromise future) {
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        WriteTimeoutTask task = lastTask;
+        lastTask = null;
+        while (task != null) {
+            task.scheduledFuture.cancel(false);
+            WriteTimeoutTask prev = task.prev;
+            task.prev = null;
+            task.next = null;
+            task = prev;
+        }
+    }
+
+    private void scheduleTimeout(final ChannelHandlerContext ctx, final ChannelPromise promise) {
         // Schedule a timeout.
-        final ScheduledFuture<?> sf = ctx.executor().schedule(new OneTimeTask() {
-            @Override
-            public void run() {
-                // Was not written yet so issue a write timeout
-                // The future itself will be failed with a ClosedChannelException once the close() was issued
-                // See https://github.com/netty/netty/issues/2159
-                if (!future.isDone()) {
-                    try {
-                        writeTimedOut(ctx);
-                    } catch (Throwable t) {
-                        ctx.fireExceptionCaught(t);
-                    }
+        final WriteTimeoutTask task = new WriteTimeoutTask(ctx, promise);
```

Comment:
Can we also have WriteTimeoutTask implement ChannelFutureListener and so safe one object allocation ?